openedx/features/journals/api.py

```diff
@@ -0,0 +1,276 @@
+
+    Example:
+        >>> get_cache_key(site_domain="example.com", resource="enterprise-learner")
+        # Here is key format for above call
+        # "site_domain:example.com__resource:enterprise-learner"
+        a54349175618ff1659dee0978e3149ca
+
+    Arguments:
+        **kwargs: Key word arguments that need to be present in cache key.
+
+    Returns:
+         An MD5 encoded key uniquely identified by the key word arguments.
+    """
+    key = '__'.join(['{}:{}'.format(item, value) for item, value in six.iteritems(kwargs)])
+
+    return hashlib.md5(key).hexdigest()
+
+
+def journals_enabled():
+    """
+    Determines whether the Journals app is installed and enabled
+    """
+    return 'openedx.features.journals.apps.JournalsConfig' \
+        in settings.INSTALLED_APPS and settings.FEATURES.get('ENABLE_JOURNAL_INTEGRATION', False)
+
+
+def get_journals(site):
+    """Retrieve journals from the discovery service.
+
+    Keyword Arguments:
+        site (Site): Site object for the request
+        will be returned
+
+    Returns:
+        list of dict, representing journals
+    """
+    if not journals_enabled():
+        return []
+
+    api_resource = 'journals'
+    orgs = configuration_helpers.get_current_site_orgs()
+
+    cache_key = get_cache_key(
+        site_domain=site.domain,
+        resource=api_resource,
+        orgs=orgs
+    )
+
+    # look up in cache
+    journals = cache.get(cache_key)
+
+    if not journals:
+        api_client = DiscoveryApiClient()
+        if not api_client:
+            return []
+        journals = api_client.get_journals(orgs)
+        cache.set(cache_key, journals, JOURNALS_CACHE_TIMEOUT)
+
+    return journals
+
+
+def fix_course_images(bundle):
+    """
+        Set the image for a course. If the course has an image, use that. Otherwise use the first
+        course run that has an image.
+    """
+    for course in bundle['courses']:
+        course_image = course['image'].get('src') if course.get('image') else None
+        if course_image:
+            # Course already had image and we don't need to check course runs
+            continue
+        for course_run in course['course_runs']:
+            if course_run['image']:
+                course['image'] = course_run['image']
+                break
+
+
+def get_journal_bundles(site, bundle_uuid=''):
+    """Retrieve journal bundles from the discovery service.
+
+    Returns:
+        list of dict, representing journal bundles
+    """
+    if not journals_enabled():
+        return []
+
+    api_resource = 'journal_bundles'
+
+    cache_key = get_cache_key(
+        site_domain=site.domain,
+        resource=api_resource,
+        bundle_uuid=bundle_uuid
+    )
+
+    journal_bundles = cache.get(cache_key)
+
+    if not journal_bundles:
+        api_client = DiscoveryApiClient()
+        if not api_client:
+            return []
+        journal_bundles = api_client.get_journal_bundles(uuid=bundle_uuid)
```

Comment:
`get_journal_bundles` can return `[]` which is falsey which essentially means we stop caching when something is going wrong (500s from discovery) or if no results is a valid response and risk flooding the discovery service with requests.

I know @BenjiLee has talked about this in some of his performance demos, but I forget if this is something handled for us now or if we still need to code around this issue.

Reply:
I believe we need to handle it ourselves per https://github.com/edx/open-edx-proposals/blob/master/oeps/oep-0022-bp-django-caches.rst

Reply:
fixed per oep proposal